// retune/dist/chunk-4FKYK3XT.js
import { scanDesignTokens, findTokensForValue } from './chunk-THD5EQWV.js';
import { finder } from '@medv/finder';
import { parse } from 'parsel-js';

// src/variables/categories.ts
var PROPERTY_CATEGORY = {
  // Spacing
  "padding": "spacing",
  "padding-top": "spacing",
  "padding-right": "spacing",
  "padding-bottom": "spacing",
  "padding-left": "spacing",
  "padding-inline": "spacing",
  "padding-inline-start": "spacing",
  "padding-inline-end": "spacing",
  "padding-block": "spacing",
  "padding-block-start": "spacing",
  "padding-block-end": "spacing",
  "margin": "spacing",
  "margin-top": "spacing",
  "margin-right": "spacing",
  "margin-bottom": "spacing",
  "margin-left": "spacing",
  "margin-inline": "spacing",
  "margin-inline-start": "spacing",
  "margin-inline-end": "spacing",
  "margin-block": "spacing",
  "margin-block-start": "spacing",
  "margin-block-end": "spacing",
  "gap": "spacing",
  "row-gap": "spacing",
  "column-gap": "spacing",
  // Sizing
  "width": "sizing",
  "height": "sizing",
  "min-width": "sizing",
  "max-width": "sizing",
  "min-height": "sizing",
  "max-height": "sizing",
  "inline-size": "sizing",
  "block-size": "sizing",
  "min-inline-size": "sizing",
  "max-inline-size": "sizing",
  "min-block-size": "sizing",
  "max-block-size": "sizing",
  // Colors
  "color": "colors",
  "background-color": "colors",
  "border-color": "colors",
  "border-top-color": "colors",
  "border-right-color": "colors",
  "border-bottom-color": "colors",
  "border-left-color": "colors",
  "border-inline-start-color": "colors",
  "border-inline-end-color": "colors",
  "border-block-start-color": "colors",
  "border-block-end-color": "colors",
  "outline-color": "colors",
  "text-decoration-color": "colors",
  "accent-color": "colors",
  "caret-color": "colors",
  "fill": "colors",
  "stroke": "colors",
  // Typography (per-property categories)
  "font-size": "font-size",
  "font-weight": "font-weight",
  "line-height": "line-height",
  "letter-spacing": "letter-spacing",
  "font-family": "font-family",
  // Border radius
  "border-radius": "border-radius",
  "border-top-left-radius": "border-radius",
  "border-top-right-radius": "border-radius",
  "border-bottom-left-radius": "border-radius",
  "border-bottom-right-radius": "border-radius",
  "border-start-start-radius": "border-radius",
  "border-start-end-radius": "border-radius",
  "border-end-start-radius": "border-radius",
  "border-end-end-radius": "border-radius",
  // Border width
  "border-width": "border-width",
  "border-top-width": "border-width",
  "border-right-width": "border-width",
  "border-bottom-width": "border-width",
  "border-left-width": "border-width",
  "border-inline-start-width": "border-width",
  "border-inline-end-width": "border-width",
  "border-block-start-width": "border-width",
  "border-block-end-width": "border-width",
  // Box shadow
  "box-shadow": "box-shadow",
  // Opacity
  "opacity": "opacity",
  // Layout
  "display": "layout",
  "flex-direction": "layout",
  "align-items": "layout",
  "justify-content": "layout",
  "flex-wrap": "layout",
  "position": "layout"
};
function getCategoryForProperty(prop) {
  return PROPERTY_CATEGORY[prop] ?? null;
}
function getSelector(element) {
  try {
    return finder(element, {
      root: document.body,
      className: (name) => {
        if (name.startsWith("_") || name.startsWith("css-")) return false;
        if (/^[a-z]{1,3}[A-Za-z0-9_]{8,}$/.test(name)) return false;
        return true;
      },
      seedMinLength: 1,
      optimizedMinLength: 2,
      threshold: 1e3
    });
  } catch {
    return buildFallbackSelector(element);
  }
}
function buildFallbackSelector(el) {
  const parts = [];
  let current = el;
  while (current && current !== document.body) {
    let part = current.tagName.toLowerCase();
    if (current.id) {
      parts.unshift(`#${current.id}`);
      break;
    }
    const parent = current.parentElement;
    if (parent) {
      const siblings = Array.from(parent.children).filter(
        (c) => c.tagName === current.tagName
      );
      if (siblings.length > 1) {
        const index = siblings.indexOf(current) + 1;
        part += `:nth-of-type(${index})`;
      }
    }
    parts.unshift(part);
    current = parent;
  }
  return parts.join(" > ");
}
function getSharedSelector(element) {
  const candidates = getSelectorCandidates(element);
  for (const c of candidates) {
    if (c.count > 1) return c;
  }
  return candidates.length > 0 ? candidates[0] : null;
}
function isHashedClass(name) {
  if (name.includes("__") || name.includes("--")) return false;
  if (name.startsWith("_") || name.startsWith("css-")) return true;
  if (/^[a-z]{1,3}[A-Za-z0-9_]{8,}$/.test(name)) return true;
  return false;
}
var UTILITY_STEMS = /* @__PURE__ */ new Set([
  // Spacing
  "m",
  "mx",
  "my",
  "mt",
  "mr",
  "mb",
  "ml",
  "ms",
  "me",
  "p",
  "px",
  "py",
  "pt",
  "pr",
  "pb",
  "pl",
  "ps",
  "pe",
  "space",
  "gap",
  // Sizing
  "w",
  "h",
  "min-w",
  "max-w",
  "min-h",
  "max-h",
  "size",
  // Typography
  "text",
  "font",
  "leading",
  "tracking",
  "indent",
  // Colors
  "bg",
  "from",
  "via",
  "to",
  // Borders
  "border",
  "rounded",
  "ring",
  "outline",
  "divide",
  // Layout
  "flex",
  "grid",
  "col",
  "row",
  "place",
  "items",
  "justify",
  "self",
  "order",
  "float",
  "clear",
  "basis",
  "grow",
  "shrink",
  "wrap",
  // Positioning
  "inset",
  "top",
  "right",
  "bottom",
  "left",
  "z",
  // Visual
  "shadow",
  "opacity",
  "blur",
  "brightness",
  "contrast",
  "saturate",
  "backdrop",
  "mix-blend",
  // Behavior
  "overflow",
  "overscroll",
  "scroll",
  "snap",
  "touch",
  "select",
  "cursor",
  "pointer-events",
  "resize",
  "appearance",
  // Transitions
  "transition",
  "duration",
  "ease",
  "delay",
  "animate",
  // Transforms
  "scale",
  "rotate",
  "translate",
  "skew",
  "origin",
  // Other
  "aspect",
  "columns",
  "break",
  "object",
  "decoration",
  "underline",
  "overline",
  "line-through",
  "no-underline",
  "uppercase",
  "lowercase",
  "capitalize",
  "normal-case",
  "truncate",
  "antialiased",
  "subpixel-antialiased",
  "whitespace",
  "hyphens",
  "content",
  "list",
  "sr",
  "not-sr",
  "will-change",
  "fill",
  "stroke",
  "caret",
  "accent",
  "container",
  "prose",
  // Single-word Tailwind utilities
  "block",
  "inline",
  "inline-block",
  "inline-flex",
  "inline-grid",
  "hidden",
  "visible",
  "invisible",
  "collapse",
  "static",
  "relative",
  "absolute",
  "fixed",
  "sticky",
  "isolate",
  "isolation",
  "table",
  "table-caption",
  "table-cell",
  "table-column",
  "table-row",
  "table-footer-group",
  "table-header-group",
  "table-row-group",
  "table-column-group",
  "contents",
  "flow-root",
  "line-clamp",
  "italic",
  "not-italic",
  // Tailwind v4 additions
  "inset-ring",
  "inset-shadow",
  "field-sizing",
  "text-wrap",
  "text-nowrap",
  "text-balance",
  "text-pretty"
]);
var SEMANTIC_STEMS = /* @__PURE__ */ new Set([
  "btn",
  "button",
  "card",
  "modal",
  "nav",
  "header",
  "footer",
  "sidebar",
  "hero",
  "section",
  "container-fluid",
  "wrapper",
  "layout",
  "page",
  "view",
  "panel",
  "dialog",
  "menu",
  "toolbar",
  "badge",
  "chip",
  "avatar",
  "icon",
  "logo",
  "form",
  "input",
  "item",
  "link",
  "tab",
  "accordion",
  "carousel",
  "dropdown",
  "tooltip",
  "popover",
  "alert",
  "toast",
  "banner",
  "widget"
]);
var UTILITY_KEYWORD_VALUES = /* @__PURE__ */ new Set([
  // font-weight
  "thin",
  "extralight",
  "light",
  "normal",
  "medium",
  "semibold",
  "bold",
  "extrabold",
  "black",
  // leading
  "none",
  "tight",
  "snug",
  "relaxed",
  "loose",
  // tracking
  "tighter",
  "wide",
  "wider",
  "widest",
  // overflow / visibility
  "hidden",
  "visible",
  "scroll",
  "auto",
  "clip",
  // shadow
  "inner",
  "outer",
  "inset",
  // text-align / justify / align
  "left",
  "center",
  "right",
  "justify",
  "start",
  "end",
  // flex direction / wrap
  "row",
  "col",
  "wrap",
  "nowrap",
  "reverse",
  // object-fit
  "cover",
  "contain",
  "fill",
  // alignment
  "baseline",
  "stretch",
  "between",
  "around",
  "evenly",
  // cursor
  "pointer",
  "default",
  "move",
  "grab",
  "grabbing",
  "not-allowed",
  "wait",
  "crosshair",
  // misc
  "ellipsis",
  "truncate",
  "clamp"
]);
var VARIANT_PREFIX = /^(?:sm|md|lg|xl|2xl|3xl|4xl|5xl|min-sm|min-md|min-lg|min-xl|min-2xl|max-sm|max-md|max-lg|max-xl|max-2xl|dark|light|hover|focus|focus-within|focus-visible|active|visited|target|first|last|only|odd|even|first-of-type|last-of-type|only-of-type|empty|disabled|enabled|checked|indeterminate|default|required|valid|invalid|in-range|out-of-range|placeholder-shown|autofill|read-only|open|closed|before|after|first-letter|first-line|marker|selection|file|backdrop|placeholder|group-hover|group-focus|group-active|group-first|group-last|peer-hover|peer-focus|peer-checked|peer-disabled|has|not|is|where|supports|aria|data|rtl|ltr|print|portrait|landscape|motion-safe|motion-reduce|contrast-more|contrast-less|forced-colors):/;
function extractStem(name) {
  const stripped = name.replace(/^.*?:/, "");
  const clean = stripped.replace(/^[-!]/, "");
  const multiMatch = clean.match(/^([a-z]+-[a-z]+)(?=-|$)/);
  if (multiMatch && UTILITY_STEMS.has(multiMatch[1])) return multiMatch[1];
  return clean.split("-")[0];
}
function scoreNamePattern(name) {
  if (!name || !name.trim()) return { score: 0, confidence: 1 };
  if (VARIANT_PREFIX.test(name)) return { score: 1, confidence: 0.95 };
  if (/\[.+\]/.test(name)) return { score: 1, confidence: 0.95 };
  if (/^[a-z][\w-]*\/\d+$/.test(name)) return { score: 1, confidence: 0.95 };
  if (/!$/.test(name)) return { score: 0.95, confidence: 0.9 };
  const stem = extractStem(name);
  if (name.includes("__") || /--[a-z]/.test(name)) return { score: 0.05, confidence: 0.8 };
  if (UTILITY_STEMS.has(name)) {
    return { score: 0.8, confidence: 0.75 };
  }
  if (SEMANTIC_STEMS.has(stem)) return { score: 0.15, confidence: 0.7 };
  if (/^(is-|has-|not-|no-|active|disabled|selected|open|closed|collapsed|expanded|loading|error|success|warning|checked|focused|pressed|dragging)$/.test(name)) {
    return { score: 0.25, confidence: 0.5 };
  }
  const isUtilityStem = UTILITY_STEMS.has(stem);
  const hasValueSuffix = /-(\d+\.?\d*|xs|sm|md|lg|xl|2xl|3xl|4xl|5xl|6xl|7xl|auto|full|screen|fit|min|max|none|px|0\.5|1\.5|2\.5|3\.5|\d+\/\d+)$/.test(name);
  const hasColorPattern = /-(slate|gray|zinc|neutral|stone|red|orange|amber|yellow|lime|green|emerald|teal|cyan|sky|blue|indigo|violet|purple|fuchsia|pink|rose|black|white|transparent|current|inherit)-?\d*$/.test(name);
  const suffix = name.length > stem.length ? name.slice(stem.length + 1) : "";
  const hasKeywordValue = suffix !== "" && UTILITY_KEYWORD_VALUES.has(suffix);
  if (isUtilityStem && (hasValueSuffix || hasColorPattern || hasKeywordValue)) {
    return { score: 0.85, confidence: 0.8 };
  }
  if (isUtilityStem) {
    return { score: 0.5, confidence: 0.4 };
  }
  if (hasValueSuffix && !isUtilityStem) {
    return { score: 0.55, confidence: 0.35 };
  }
  if (name.length <= 6 && /^[a-z]+-\d/.test(name)) {
    return { score: 0.7, confidence: 0.5 };
  }
  return { score: 0.35, confidence: 0.2 };
}
function getPropertyFamily(prop) {
  if (prop.startsWith("padding-")) return "padding";
  if (prop.startsWith("margin-")) return "margin";
  if (prop.endsWith("-radius") && prop.startsWith("border-")) return "border-radius";
  if (/^border-(top|right|bottom|left)-(width|style|color)$/.test(prop)) return "border";
  if (prop.startsWith("border-") && (prop.startsWith("border-block") || prop.startsWith("border-inline"))) return "border";
  if (prop.startsWith("background-")) return "background";
  if (prop.startsWith("overflow-")) return "overflow";
  if (prop === "column-gap" || prop === "row-gap") return "gap";
  if (prop.startsWith("outline-") && prop !== "outline-offset") return "outline";
  if (prop.startsWith("text-decoration-")) return "text-decoration";
  if (prop.startsWith("transition-")) return "transition";
  if (prop.startsWith("animation-") && !prop.startsWith("animation-range")) return "animation";
  if (prop.startsWith("scroll-margin-")) return "scroll-margin";
  if (prop.startsWith("scroll-padding-")) return "scroll-padding";
  if (prop.startsWith("list-style-")) return "list-style";
  if (prop.startsWith("grid-template-")) return "grid-template";
  if (prop.startsWith("grid-auto-")) return "grid-auto";
  if (prop === "flex-grow" || prop === "flex-shrink" || prop === "flex-basis") return "flex";
  if (prop === "column-count" || prop === "column-width") return "columns";
  if (prop.startsWith("place-")) return prop;
  return prop;
}
function countAuthoredProperties(style) {
  const families = /* @__PURE__ */ new Set();
  for (let i = 0; i < style.length; i++) {
    const prop = style[i];
    if (prop.startsWith("--")) continue;
    families.add(getPropertyFamily(prop));
  }
  return families.size;
}
function isSimpleClassSelector(selectorText) {
  return selectorText.split(",").every((sel) => {
    const s = sel.trim();
    const noPseudo = s.replace(/::?[\w-]+(?:\(.*?\))?/g, "");
    const cleaned = noPseudo.replace(/\[.*?\]/g, "").trim();
    if (/[\s>+~]/.test(cleaned)) return false;
    const classes = cleaned.match(/\./g);
    return classes !== null && classes.length === 1;
  });
}
function analyzeElementClasses(element) {
  const elClasses = element.classList ? Array.from(element.classList) : [];
  const result = {
    compoundClasses: /* @__PURE__ */ new Set(),
    utilityClasses: /* @__PURE__ */ new Set(),
    classesFoundInRules: /* @__PURE__ */ new Set(),
    scores: /* @__PURE__ */ new Map()
  };
  if (elClasses.length === 0) return result;
  const maxPropCount = /* @__PURE__ */ new Map();
  const hasComplexSelector = /* @__PURE__ */ new Set();
  for (const sheet of document.styleSheets) {
    let rules;
    try {
      rules = sheet.cssRules;
    } catch {
      continue;
    }
    walkRules(rules, elClasses, element, result, maxPropCount, hasComplexSelector, null);
  }
  const nonHashed = elClasses.filter((c) => !isHashedClass(c));
  for (const cls of nonHashed) {
    if (result.utilityClasses.has(cls)) {
      result.scores.set(cls, { score: 1, confidence: 0.98, verdict: "utility" });
      continue;
    }
    const nameSignal = scoreNamePattern(cls);
    const foundInRules = result.classesFoundInRules.has(cls);
    let sheetScore = null;
    if (foundInRules) {
      const maxProps = maxPropCount.get(cls) ?? 0;
      const isComplex = hasComplexSelector.has(cls);
      if (maxProps <= 1 && !isComplex) {
        sheetScore = { score: 0.9, confidence: 0.85 };
      } else if (maxProps <= 2 && !isComplex) {
        sheetScore = { score: 0.75, confidence: 0.8 };
      } else if (maxProps <= 3 && !isComplex) {
        sheetScore = { score: 0.5, confidence: 0.65 };
      } else if (maxProps >= 5) {
        sheetScore = { score: 0.1, confidence: 0.85 };
      } else if (isComplex) {
        sheetScore = { score: Math.max(0, 0.3 - maxProps * 0.05), confidence: 0.75 };
      } else {
        sheetScore = { score: 0.4, confidence: 0.6 };
      }
    }
    let finalScore;
    let finalConfidence;
    if (sheetScore) {
      if (nameSignal.score <= 0.25 && sheetScore.score >= 0.65) {
        finalScore = nameSignal.score;
        finalConfidence = nameSignal.confidence;
      } else {
        finalScore = sheetScore.score * 0.65 + nameSignal.score * 0.35;
        finalConfidence = Math.max(sheetScore.confidence, nameSignal.confidence);
        if (Math.abs(sheetScore.score - nameSignal.score) < 0.2) {
          finalConfidence = Math.min(finalConfidence + 0.1, 1);
        }
      }
    } else {
      finalScore = nameSignal.score;
      finalConfidence = nameSignal.confidence * 0.9;
    }
    if (nonHashed.length >= 10 && finalScore >= 0.35 && finalScore <= 0.65) {
      finalScore = Math.min(finalScore + 0.1, 1);
    } else if (nonHashed.length <= 2 && finalScore >= 0.35 && finalScore <= 0.65) {
      finalScore = Math.max(finalScore - 0.1, 0);
    }
    let verdict;
    if (finalScore >= 0.65) {
      verdict = "utility";
      result.utilityClasses.add(cls);
    } else if (finalScore <= 0.35) {
      verdict = "semantic";
    } else if (finalConfidence >= 0.7) {
      verdict = finalScore >= 0.5 ? "utility" : "semantic";
      if (verdict === "utility") result.utilityClasses.add(cls);
    } else {
      verdict = "ambiguous";
    }
    result.scores.set(cls, { score: finalScore, confidence: finalConfidence, verdict });
  }
  return result;
}
function walkRules(rules, elClasses, element, result, maxPropCount, hasComplexSelector, layerName) {
  for (let i = 0; i < rules.length; i++) {
    const rule = rules[i];
    if (typeof CSSLayerBlockRule !== "undefined" && rule instanceof CSSLayerBlockRule) {
      walkRules(rule.cssRules, elClasses, element, result, maxPropCount, hasComplexSelector, rule.name);
      continue;
    }
    if (rule instanceof CSSGroupingRule) {
      walkRules(rule.cssRules, elClasses, element, result, maxPropCount, hasComplexSelector, layerName);
      continue;
    }
    if (!(rule instanceof CSSStyleRule)) continue;
    const sel = rule.selectorText;
    const selectorClasses = sel.match(/\.([a-zA-Z0-9_-]+)/g)?.map((c) => c.slice(1)) || [];
    const overlap = selectorClasses.filter((c) => elClasses.includes(c));
    if (overlap.length === 0) continue;
    try {
      if (!element.matches(sel)) continue;
    } catch {
      continue;
    }
    for (const c of overlap) result.classesFoundInRules.add(c);
    if (overlap.length >= 2) {
      for (const c of overlap) result.compoundClasses.add(c);
    }
    if (layerName === "utilities") {
      for (const c of overlap) result.utilityClasses.add(c);
      continue;
    }
    const propCount = countAuthoredProperties(rule.style);
    const simple = isSimpleClassSelector(sel);
    for (const c of overlap) {
      const prev = maxPropCount.get(c) ?? 0;
      maxPropCount.set(c, Math.max(prev, propCount));
      if (!simple) hasComplexSelector.add(c);
    }
  }
}
function getSelectorCandidates(element) {
  const el = element;
  if (!el.classList || el.classList.length === 0) return [];
  const analysis = analyzeElementClasses(element);
  const nonHashed = Array.from(el.classList).filter((name) => !isHashedClass(name));
  if (nonHashed.length === 0) return [];
  const candidates = [];
  const semantic = [];
  const ambiguous = [];
  const utility = [];
  for (const name of nonHashed) {
    const info = analysis.scores.get(name);
    if (!info) {
      ambiguous.push(name);
      continue;
    }
    if (analysis.compoundClasses.has(name) && info.verdict === "utility") {
      semantic.push(name);
      continue;
    }
    switch (info.verdict) {
      case "semantic":
        semantic.push(name);
        break;
      case "ambiguous":
        ambiguous.push(name);
        break;
      case "utility":
        utility.push(name);
        break;
    }
  }
  for (const c of semantic) {
    const selector = `.${CSS.escape(c)}`;
    try {
      candidates.push({
        selector,
        count: document.querySelectorAll(selector).length,
        verdict: "semantic"
      });
    } catch {
    }
  }
  for (const c of ambiguous) {
    const selector = `.${CSS.escape(c)}`;
    try {
      candidates.push({
        selector,
        count: document.querySelectorAll(selector).length,
        verdict: "ambiguous"
      });
    } catch {
    }
  }
  for (const c of utility) {
    const selector = `.${CSS.escape(c)}`;
    try {
      candidates.push({
        selector,
        count: document.querySelectorAll(selector).length,
        verdict: "utility"
      });
    } catch {
    }
  }
  const verdictPriority = { semantic: 0, ambiguous: 1, utility: 2 };
  candidates.sort((a, b) => {
    const vp = verdictPriority[a.verdict] - verdictPriority[b.verdict];
    if (vp !== 0) return vp;
    return b.count - a.count;
  });
  return candidates;
}
function getFiber(element) {
  const key = Object.keys(element).find((k) => k.startsWith("__reactFiber$"));
  return key ? element[key] : null;
}
function getReactComponentHierarchy(element) {
  const fiber = getFiber(element);
  if (!fiber) return [];
  const components = [];
  let current = fiber;
  while (current) {
    if (typeof current.type === "function" || typeof current.type === "object") {
      const name = current.type?.displayName || current.type?.name || current.elementType?.displayName || current.elementType?.name;
      if (name && !isFrameworkInternal(name)) {
        components.push(name);
      }
    }
    current = current.return;
  }
  return components;
}
function getDirectReactComponent(element) {
  const fiber = getFiber(element);
  if (!fiber) return null;
  let current = fiber.return;
  while (current) {
    if (typeof current.type === "function" || typeof current.type === "object") {
      const name = current.type?.displayName || current.type?.name || current.elementType?.displayName || current.elementType?.name;
      if (name && !isFrameworkInternal(name)) {
        const rootDom = findFirstDomFiber(current);
        return rootDom === element ? name : null;
      }
    }
    current = current.return;
  }
  return null;
}
function getDirectComponentFiber(element) {
  const fiber = getFiber(element);
  if (!fiber) return null;
  let current = fiber.return;
  while (current) {
    if (typeof current.type === "function" || typeof current.type === "object") {
      const name = current.type?.displayName || current.type?.name || current.elementType?.displayName || current.elementType?.name;
      if (name && !isFrameworkInternal(name)) {
        const rootDom = findFirstDomFiber(current);
        return rootDom === element ? current : null;
      }
    }
    current = current.return;
  }
  return null;
}
function getDirectReactProps(element) {
  const fiber = getDirectComponentFiber(element);
  if (!fiber?.memoizedProps) return null;
  const raw = fiber.memoizedProps;
  const props = {};
  for (const key of Object.keys(raw)) {
    if (key === "children" || key === "ref" || key === "key" || key === "params" || key === "searchParams") continue;
    try {
      props[key] = raw[key];
    } catch {
    }
  }
  return Object.keys(props).length > 0 ? props : null;
}
function getDirectReactState(element) {
  const fiber = getDirectComponentFiber(element);
  if (!fiber) return null;
  const hooks = [];
  let hook = fiber.memoizedState;
  let index = 0;
  while (hook) {
    const type = identifyHookType(hook);
    const hasDispatch = !!(hook.queue && typeof hook.queue.dispatch === "function");
    if (type === "state" || type === "reducer" || type === "ref") {
      hooks.push({
        index,
        type,
        value: type === "ref" ? hook.memoizedState?.current : hook.memoizedState,
        hasDispatch
      });
    }
    hook = hook.next;
    index++;
  }
  return hooks.length > 0 ? hooks : null;
}
function findFirstDomFiber(fiber) {
  if (fiber.stateNode instanceof Element) return fiber.stateNode;
  let child = fiber.child;
  while (child) {
    const found = findFirstDomFiber(child);
    if (found) return found;
    child = child.sibling;
  }
  return null;
}
function isFrameworkInternal(name) {
  if (name.startsWith("_")) return true;
  if (/^(Fragment|Suspense|StrictMode|Profiler|Lazy|Memo|Forward)/.test(name)) return true;
  if (/(?:Provider|Consumer|Context|Boundary|Handler|Root|Wrapper)$/.test(name)) return true;
  if (/(?:Router|Layout|Template|Loading|Segment|Fallback|Reload|Manager|Metadata|Viewport|Head|Script|Link)/.test(name)) return true;
  if (/^(Home|App|Page|Main|Index|Default|View|Screen|Dashboard|Root)$/.test(name)) return true;
  if (name.length <= 2) return true;
  return false;
}
function getReactSource(element) {
  const fiber = getFiber(element);
  if (!fiber) return null;
  let current = fiber;
  while (current) {
    if (current._debugSource) {
      return {
        fileName: current._debugSource.fileName,
        lineNumber: current._debugSource.lineNumber,
        columnNumber: current._debugSource.columnNumber
      };
    }
    if (current._debugOwner?._debugSource) {
      return {
        fileName: current._debugOwner._debugSource.fileName,
        lineNumber: current._debugOwner._debugSource.lineNumber,
        columnNumber: current._debugOwner._debugSource.columnNumber
      };
    }
    current = current.return;
  }
  return null;
}
function identifyHookType(hook) {
  const state = hook.memoizedState;
  const queue = hook.queue;
  if (queue && typeof queue.dispatch === "function") {
    return queue.lastRenderedReducer?.name === "basicStateReducer" ? "state" : "reducer";
  }
  if (state !== null && typeof state === "object" && "current" in state && Object.keys(state).length === 1) {
    return "ref";
  }
  if (state !== null && typeof state === "object" && "tag" in state && "create" in state && "deps" in state) {
    return "effect";
  }
  if (Array.isArray(state) && state.length === 2 && Array.isArray(state[1])) {
    return "memo";
  }
  return "unknown";
}
function setReactState(element, hookIndex, newValue) {
  const fiber = getFiber(element);
  if (!fiber) return false;
  let componentFiber = fiber;
  while (componentFiber) {
    if ((typeof componentFiber.type === "function" || typeof componentFiber.type === "object") && !isFrameworkInternal(componentFiber.type?.displayName || componentFiber.type?.name || "")) {
      break;
    }
    componentFiber = componentFiber.return;
  }
  if (!componentFiber) return false;
  let hook = componentFiber.memoizedState;
  for (let i = 0; i < hookIndex && hook; i++) {
    hook = hook.next;
  }
  if (!hook?.queue?.dispatch) return false;
  hook.queue.dispatch(newValue);
  return true;
}
var ARIA_LABELS = {
  "aria-expanded": { "true": "Expanded", "false": "Collapsed" },
  "aria-selected": { "true": "Selected", "false": "Deselected" },
  "aria-disabled": { "true": "Disabled" },
  "aria-checked": { "true": "Checked", "mixed": "Partially Checked" },
  "aria-pressed": { "true": "Pressed", "mixed": "Partially Pressed" },
  "aria-current": { "page": "Current Page", "step": "Current Step", "true": "Current" },
  "aria-hidden": { "true": "Hidden" },
  "aria-busy": { "true": "Loading" },
  "aria-invalid": { "true": "Invalid", "grammar": "Grammar Error", "spelling": "Spelling Error" },
  "aria-required": { "true": "Required" },
  "aria-modal": { "true": "Modal" }
};
function humanizeAncestorPart(ancestor) {
  const ariaMatch = ancestor.match(/\[aria-([\w-]+)(?:="([^"]*)")?\]/);
  if (ariaMatch) {
    const [, attr, val] = ariaMatch;
    const key = `aria-${attr}`;
    if (ARIA_LABELS[key]) {
      return ARIA_LABELS[key][val || "true"] || titleCase(val || attr);
    }
    return titleCase(val || attr);
  }
  const dataValMatch = ancestor.match(/\[data-[\w-]+=["']?([^"'\]]+)["']?\]/);
  if (dataValMatch) return titleCase(dataValMatch[1]);
  const dataBoolMatch = ancestor.match(/\[data-([\w-]+)\]/);
  if (dataBoolMatch) return titleCase(dataBoolMatch[1]);
  const bemMatch = ancestor.match(/\.([\w-]+)--([\w-]+)/);
  if (bemMatch) return titleCase(bemMatch[2]);
  const stateMatch = ancestor.match(/\.(is|has)-([\w-]+)/);
  if (stateMatch) return titleCase(stateMatch[2]);
  const classMatch = ancestor.match(/\.([\w-]+)/);
  if (classMatch) return titleCase(classMatch[1]);
  return ancestor.replace(/[.\[\]="']/g, " ").trim();
}
function titleCase(s) {
  return s.split("-").map((w) => w.charAt(0).toUpperCase() + w.slice(1)).join(" ");
}
function isNoiseAncestor(selectorText) {
  if (/\[data-v-[a-f0-9]+\]/.test(selectorText)) return true;
  if (/\[_ng(host|content)-[\w-]+\]/.test(selectorText)) return true;
  if (/\.(css|sc|emotion)-[a-zA-Z0-9]{4,}/.test(selectorText) && !/[A-Z]/.test(selectorText)) return true;
  return false;
}
function splitSelector(selectorText) {
  let ast;
  try {
    ast = parse(selectorText, { recursive: true });
  } catch {
    return null;
  }
  if (!ast || ast.type !== "complex") return null;
  const rightContent = ast.right?.content;
  if (!rightContent) return null;
  const full = selectorText.trim();
  const rightIdx = full.lastIndexOf(rightContent);
  if (rightIdx <= 0) return null;
  const ancestorRaw = full.slice(0, rightIdx).trim();
  const ancestorPart = ancestorRaw.replace(/[>+~\s]+$/, "").trim();
  if (!ancestorPart) return null;
  return { ancestorPart, elementPart: rightContent };
}
function elementPartMatchesElement(elementPart, element) {
  const partClasses = elementPart.match(/\.([a-zA-Z0-9_-]+)/g)?.map((c) => c.slice(1)) || [];
  if (partClasses.length === 0) {
    const tag = elementPart.replace(/[:\[].*/g, "").trim().toLowerCase();
    return tag === element.tagName.toLowerCase();
  }
  return partClasses.some((c) => element.classList.contains(c));
}
function getAncestorScopes(element) {
  const seen = /* @__PURE__ */ new Map();
  for (const sheet of document.styleSheets) {
    let rules;
    try {
      rules = sheet.cssRules;
    } catch {
      continue;
    }
    walkRulesForAncestors(rules, element, seen);
  }
  const result = Array.from(seen.values());
  result.sort((a, b) => b.count - a.count);
  return result;
}
function walkRulesForAncestors(rules, element, seen) {
  for (let i = 0; i < rules.length; i++) {
    const rule = rules[i];
    if (rule instanceof CSSGroupingRule || typeof CSSLayerBlockRule !== "undefined" && rule instanceof CSSLayerBlockRule) {
      walkRulesForAncestors(rule.cssRules, element, seen);
      continue;
    }
    if (!(rule instanceof CSSStyleRule)) continue;
    const sel = rule.selectorText;
    if (sel.includes(":hover") || sel.includes(":focus") || sel.includes(":active")) continue;
    try {
      if (!element.matches(sel)) continue;
    } catch {
      continue;
    }
    for (const part of sel.split(",")) {
      const trimmed = part.trim();
      const noParen = trimmed.replace(/\([^)]*\)/g, "");
      if (!/[\s>]/.test(noParen)) continue;
      const split = splitSelector(trimmed);
      if (!split) continue;
      if (!elementPartMatchesElement(split.elementPart, element)) continue;
      if (isNoiseAncestor(split.ancestorPart)) continue;
      const fullSelector = trimmed;
      if (seen.has(fullSelector)) continue;
      let count;
      try {
        count = document.querySelectorAll(fullSelector).length;
      } catch {
        continue;
      }
      if (count <= 0) continue;
      seen.set(fullSelector, {
        fullSelector,
        ancestorPart: split.ancestorPart,
        label: humanizeAncestorPart(split.ancestorPart),
        count
      });
    }
  }
}

// src/variables/registry.ts
var MAX_UTILITY_PROPS = 1;
var cachedRegistry = null;
var cachedSheetCount = -1;
function getVariableRegistry() {
  const sheetCount = document.styleSheets.length;
  if (cachedRegistry && cachedSheetCount === sheetCount) return cachedRegistry;
  cachedRegistry = buildRegistry();
  cachedSheetCount = sheetCount;
  return cachedRegistry;
}
function buildRegistry() {
  const groups = /* @__PURE__ */ new Map();
  const valueLookup = /* @__PURE__ */ new Map();
  const classLookup = /* @__PURE__ */ new Map();
  const probe = document.createElement("div");
  probe.style.cssText = "position:fixed;top:-9999px;left:-9999px;visibility:hidden;pointer-events:none;";
  document.body.appendChild(probe);
  let hasTwVars = false;
  try {
    for (const sheet of document.styleSheets) {
      let rules;
      try {
        rules = sheet.cssRules;
      } catch {
        continue;
      }
      if (!hasTwVars) hasTwVars = detectTailwindVars(rules);
      scanRules(rules, probe, groups, valueLookup, classLookup, void 0);
    }
  } finally {
    probe.remove();
  }
  for (const [category, tokens] of groups) {
    groups.set(category, sortTokens(tokens, category));
  }
  const framework = detectFramework(hasTwVars, classLookup);
  return { groups, valueLookup, classLookup, framework };
}
function detectTailwindVars(rules) {
  for (let i = 0; i < rules.length; i++) {
    const rule = rules[i];
    if (rule instanceof CSSGroupingRule || typeof CSSLayerBlockRule !== "undefined" && rule instanceof CSSLayerBlockRule) {
      if (detectTailwindVars(rule.cssRules)) return true;
      continue;
    }
    if (!(rule instanceof CSSStyleRule)) continue;
    for (let j = 0; j < rule.style.length; j++) {
      if (rule.style[j].startsWith("--tw-")) return true;
    }
  }
  return false;
}
function detectFramework(hasTwVars, classLookup) {
  if (hasTwVars) return "tailwind";
  const twPattern = /^(p|m|w|h|gap|text|bg|border|rounded|shadow|opacity|font|leading|tracking|flex|grid|z|inset|top|right|bottom|left)([xytrbl])?-/;
  let twCount = 0;
  for (const name of classLookup.keys()) {
    if (twPattern.test(name)) twCount++;
  }
  if (twCount > 10) return "tailwind";
  if (classLookup.size > 0) return "custom";
  return "unknown";
}
function scanRules(rules, probe, groups, valueLookup, classLookup, layerName) {
  for (let i = 0; i < rules.length; i++) {
    const rule = rules[i];
    if (rule instanceof CSSGroupingRule || typeof CSSLayerBlockRule !== "undefined" && rule instanceof CSSLayerBlockRule) {
      const childLayer = typeof CSSLayerBlockRule !== "undefined" && rule instanceof CSSLayerBlockRule ? rule.name || layerName : layerName;
      scanRules(rule.cssRules, probe, groups, valueLookup, classLookup, childLayer);
      continue;
    }
    if (!(rule instanceof CSSStyleRule)) continue;
    if (!isSimpleClassSelector(rule.selectorText)) continue;
    const propCount = countAuthoredProperties(rule.style);
    if (propCount === 0 || propCount > MAX_UTILITY_PROPS) continue;
    const classMatch = rule.selectorText.match(/\.([a-zA-Z0-9_-]+)/);
    if (!classMatch) continue;
    const className = classMatch[1];
    const { score } = scoreNamePattern(className);
    if (score < 0.2) continue;
    if (classLookup.has(className)) continue;
    const values = {};
    for (let j = 0; j < rule.style.length; j++) {
      const prop = rule.style[j];
      if (prop.startsWith("--")) continue;
      let val = rule.style.getPropertyValue(prop).trim();
      if (val.includes("var(")) {
        probe.style.cssText = `position:fixed;top:-9999px;left:-9999px;visibility:hidden;pointer-events:none;`;
        probe.style.setProperty(prop, val);
        val = getComputedStyle(probe).getPropertyValue(prop).trim();
        probe.style.removeProperty(prop);
      }
      if (val) values[prop] = val;
    }
    if (Object.keys(values).length === 0) continue;
    let category = null;
    let multiCategory = false;
    for (const prop of Object.keys(values)) {
      const cat = getCategoryForProperty(prop);
      if (!cat) continue;
      if (!category) {
        category = cat;
      } else if (cat !== category) {
        multiCategory = true;
        break;
      }
    }
    if (!category || multiCategory) continue;
    const token = { className, values, layerName };
    classLookup.set(className, token);
    if (!groups.has(category)) groups.set(category, []);
    groups.get(category).push(token);
    for (const [prop, val] of Object.entries(values)) {
      const key = `${prop}:${val}`;
      if (!valueLookup.has(key)) valueLookup.set(key, []);
      valueLookup.get(key).push(token);
    }
  }
}
function sortTokens(tokens, category) {
  if (category === "colors") {
    return tokens.sort((a, b) => {
      const hueA = extractHue(Object.values(a.values)[0]);
      const hueB = extractHue(Object.values(b.values)[0]);
      return hueA - hueB;
    });
  }
  return tokens.sort((a, b) => {
    const numA = parseFloat(Object.values(a.values)[0]);
    const numB = parseFloat(Object.values(b.values)[0]);
    if (isNaN(numA) && isNaN(numB)) return a.className.localeCompare(b.className);
    if (isNaN(numA)) return 1;
    if (isNaN(numB)) return -1;
    return numA - numB;
  });
}
function extractHue(color) {
  if (!color) return 0;
  const rgbMatch = color.match(/rgba?\((\d+),\s*(\d+),\s*(\d+)/);
  if (!rgbMatch) return 0;
  const r = parseInt(rgbMatch[1]) / 255;
  const g = parseInt(rgbMatch[2]) / 255;
  const b = parseInt(rgbMatch[3]) / 255;
  const max = Math.max(r, g, b), min = Math.min(r, g, b);
  if (max === min) return 0;
  const d = max - min;
  let h = 0;
  if (max === r) h = ((g - b) / d + (g < b ? 6 : 0)) / 6;
  else if (max === g) h = ((b - r) / d + 2) / 6;
  else h = ((r - g) / d + 4) / 6;
  return h * 360;
}

// src/variables/resolver.ts
var SPACE_RGB_RE = /^\s*(\d{1,3})\s+(\d{1,3})\s+(\d{1,3})\s*$/;
function isSpaceSeparatedRgb(value) {
  const m = SPACE_RGB_RE.exec(value.trim());
  if (!m) return false;
  return [m[1], m[2], m[3]].every((n) => {
    const v = parseInt(n, 10);
    return v >= 0 && v <= 255;
  });
}
function normalizeColorValue(value) {
  if (isSpaceSeparatedRgb(value)) return `rgb(${value.trim().replace(/\s+/g, ", ")})`;
  return value;
}
var FRAMEWORK_INTERNAL_PREFIXES = ["--tw-", "--chakra-", "--mantine-", "--radix-", "--nextui-"];
function resolveVariablesForElement(element, computedStyles, scopeSelector) {
  const registry = getVariableRegistry();
  const matches = /* @__PURE__ */ new Map();
  const classes = element.classList ? Array.from(element.classList) : [];
  if (classes.length > 0) {
    const activeTokens = [];
    for (const cls of classes) {
      const token = registry.classLookup.get(cls);
      if (token) activeTokens.push(token);
    }
    for (const token of activeTokens) {
      for (const [prop, tokenValue] of Object.entries(token.values)) {
        const camelProp = kebabToCamel(prop);
        const computed = computedStyles[camelProp] || computedStyles[prop];
        if (!computed) continue;
        const normalizedToken = normalizeValue(tokenValue);
        const normalizedComputed = normalizeValue(computed);
        if (normalizedToken === normalizedComputed) {
          matches.set(prop, { variable: token, property: prop });
        }
      }
    }
  }
  if (classes.length > 0) {
    const { manifestClassLookup } = getCssVariables();
    for (const cls of classes) {
      const token = manifestClassLookup.get(cls);
      if (!token) continue;
      const tokenValue = normalizeValue(Object.values(token.values)[0] || "");
      for (const [prop, computed] of Object.entries(computedStyles)) {
        const kebab = camelToKebab(prop);
        if (matches.has(kebab)) continue;
        if (normalizeValue(computed) === tokenValue) {
          matches.set(kebab, { variable: token, property: kebab });
        }
      }
    }
  }
  resolveVarReferences(element, matches, scopeSelector);
  return matches;
}
var VAR_REF_RE = /var\((--[a-zA-Z0-9_-]+)/g;
var SHORTHAND_LONGHANDS = {
  "padding": ["padding-top", "padding-right", "padding-bottom", "padding-left"],
  "margin": ["margin-top", "margin-right", "margin-bottom", "margin-left"],
  "border-radius": ["border-top-left-radius", "border-top-right-radius", "border-bottom-right-radius", "border-bottom-left-radius"],
  "gap": ["row-gap", "column-gap"],
  "border-width": ["border-top-width", "border-right-width", "border-bottom-width", "border-left-width"],
  "border-color": ["border-top-color", "border-right-color", "border-bottom-color", "border-left-color"],
  "border-style": ["border-top-style", "border-right-style", "border-bottom-style", "border-left-style"],
  "inset": ["top", "right", "bottom", "left"]
};
function resolveVarReferences(element, matches, scopeSelector) {
  const htmlEl = element;
  const { tokens: varTokens } = getCssVariables();
  if (varTokens.length === 0) return;
  const varLookup = /* @__PURE__ */ new Map();
  for (const t of varTokens) {
    const name = t.className.slice(4, -1);
    varLookup.set(name, t);
  }
  const matchSpecificity = /* @__PURE__ */ new Map();
  const tryMatch = (prop, raw, specificity = 0) => {
    if (!raw.includes("var(")) return;
    const existing = matches.get(prop);
    const prevSpec = matchSpecificity.get(prop) ?? -1;
    if (existing && !isRawUtility(existing.variable) && specificity < prevSpec) return;
    VAR_REF_RE.lastIndex = 0;
    let m;
    let variable;
    while ((m = VAR_REF_RE.exec(raw)) !== null) {
      variable = varLookup.get(m[1]);
      if (variable) break;
    }
    if (!variable) return;
    const longhands = SHORTHAND_LONGHANDS[prop];
    if (longhands) {
      for (const lh of longhands) {
        const lhExisting = matches.get(lh);
        const lhPrevSpec = matchSpecificity.get(lh) ?? -1;
        if (!lhExisting || isRawUtility(lhExisting.variable) || specificity >= lhPrevSpec) {
          matches.set(lh, { variable, property: lh });
          matchSpecificity.set(lh, specificity);
        }
      }
    } else {
      matches.set(prop, { variable, property: prop });
      matchSpecificity.set(prop, specificity);
    }
  };
  const tryClear = (prop, raw) => {
    if (!raw || raw === "") return;
    if (raw.includes("var(")) return;
    matches.delete(prop);
  };
  if (htmlEl.style && htmlEl.style.length > 0) {
    for (let i = 0; i < htmlEl.style.length; i++) {
      const prop = htmlEl.style.item(i);
      const raw = htmlEl.style.getPropertyValue(prop);
      tryMatch(prop, raw);
      tryClear(prop, raw);
    }
    for (const shorthand of Object.keys(SHORTHAND_LONGHANDS)) {
      const raw = htmlEl.style.getPropertyValue(shorthand);
      if (raw) tryMatch(shorthand, raw);
    }
  }
  const scopeClasses = scopeSelector?.match(/\.[a-zA-Z0-9_-]+/g) || null;
  try {
    for (const sheet of document.styleSheets) {
      try {
        for (const rule of sheet.cssRules) {
          if (!(rule instanceof CSSStyleRule)) continue;
          if (!element.matches(rule.selectorText)) continue;
          const ruleClasses = rule.selectorText.match(/\.[a-zA-Z0-9_-]+/g) || [];
          if (scopeClasses) {
            if (ruleClasses.length === 0) continue;
            if (!ruleClasses.every((rc) => scopeClasses.includes(rc))) continue;
          }
          const ruleSpecificity = ruleClasses.length;
          for (let i = 0; i < rule.style.length; i++) {
            const prop = rule.style.item(i);
            const raw = rule.style.getPropertyValue(prop);
            tryMatch(prop, raw, ruleSpecificity);
            tryClear(prop, raw);
          }
          for (const shorthand of Object.keys(SHORTHAND_LONGHANDS)) {
            const raw = rule.style.getPropertyValue(shorthand);
            if (raw) tryMatch(shorthand, raw, ruleSpecificity);
          }
        }
      } catch {
      }
    }
  } catch {
  }
}
function getAlternativeVariables(property, currentVariable) {
  const registry = getVariableRegistry();
  const kebab = camelToKebab(property);
  const category = getCategoryForProperty(kebab);
  if (!category) return [];
  const group = registry.groups.get(category);
  if (!group) return [];
  return group.filter((t) => {
    if (!Object.keys(t.values).includes(kebab)) return false;
    if (currentVariable && t.className === currentVariable.className) return false;
    if (isRawUtility(t)) return false;
    return true;
  });
}
function findVariableForValue(property, value) {
  const registry = getVariableRegistry();
  const kebab = camelToKebab(property);
  const key = `${kebab}:${normalizeValue(value)}`;
  const directMatches = registry.valueLookup.get(key);
  if (directMatches && directMatches.length > 0) {
    return directMatches[0];
  }
  const category = getCategoryForProperty(kebab);
  if (!category) return null;
  const group = registry.groups.get(category);
  if (!group) return null;
  const numValue = parseFloat(value);
  if (isNaN(numValue)) return null;
  let closest = null;
  let closestDist = Infinity;
  for (const token of group) {
    const tokenVal = token.values[kebab];
    if (!tokenVal) continue;
    const tokenNum = parseFloat(tokenVal);
    if (isNaN(tokenNum)) continue;
    const dist = Math.abs(tokenNum - numValue);
    if (dist < closestDist) {
      closestDist = dist;
      closest = token;
    }
  }
  if (closest && closestDist <= Math.max(2, numValue * 0.1)) {
    return closest;
  }
  return null;
}
function normalizeValue(val) {
  return val.trim().toLowerCase().replace(/\s+/g, " ");
}
function camelToKebab(str) {
  return str.replace(/[A-Z]/g, (c) => `-${c.toLowerCase()}`);
}
function kebabToCamel(str) {
  return str.replace(/-([a-z])/g, (_, c) => c.toUpperCase());
}
function isRawUtility(variable) {
  if (variable.layerName === "utilities") return true;
  if (!variable.layerName) return TW_PREFIX_LEGACY.test(variable.className);
  return false;
}
var TW_PREFIX_LEGACY = /^-?(?:p|px|py|pt|pr|pb|pl|m|mx|my|mt|mr|mb|ml|w|h|min-w|max-w|min-h|max-h|gap|space-[xy]|text|bg|border|rounded|shadow|opacity|font|leading|tracking|z|inset|top|right|bottom|left|flex|grid|grid-cols|col-span|row-span|items|justify|self|place|order|overflow|ring|outline|divide|sr|not-sr|container|aspect|columns|break|decoration|underline|overline|no-underline|cursor|transition|duration|ease|delay|animate|scale|rotate|translate|skew|origin|fill|stroke|will-change|hidden|block|inline|inline-flex|inline-block|table|absolute|relative|fixed|sticky|float|clear|isolate|object|whitespace|align|indent|truncate|uppercase|lowercase|capitalize|normal-case|italic|not-italic|antialiased|subpixel|select|resize|appearance|accent|caret|snap|touch|scroll|hyphens|content|list)($|-)/;
var cssVarCache = null;
var cssVarSheetCount = -1;
var _manifestTokens = null;
function setManifestTokens(manifest) {
  _manifestTokens = manifest;
  cssVarCache = null;
  cssVarSheetCount = -1;
}
var MANIFEST_CATEGORY_MAP = {
  spacing: "spacing",
  sizes: "sizing",
  colors: "colors",
  radii: "border-radius",
  borderWidths: "border-width",
  shadows: "box-shadow"
};
function categorizeTypographyToken(nameHint, value) {
  if (/font-weight|font_weight|font-(thin|extralight|light|normal|medium|semibold|bold|extrabold|black)/.test(nameHint)) return "font-weight";
  if (/leading|line-height|line_height/.test(nameHint)) return "line-height";
  if (/tracking|letter-spacing|letter_spacing/.test(nameHint)) return "letter-spacing";
  if (/font-family|font_family|font-heading|font-body|font-mono|font-(sans|serif|mono|centra|gelica|guardian|yahoosans)/.test(nameHint)) return "font-family";
  if (/font|text|size/.test(nameHint)) return "font-size";
  return "font-size";
}
function getManifestCategories() {
  const categories = /* @__PURE__ */ new Set();
  if (!_manifestTokens?.tokens) return categories;
  for (const categoryKey of Object.keys(_manifestTokens.tokens)) {
    const mapped = MANIFEST_CATEGORY_MAP[categoryKey];
    if (mapped) {
      categories.add(mapped);
    } else if (categoryKey === "typography") {
      categories.add("font-size");
      categories.add("font-weight");
      categories.add("line-height");
      categories.add("letter-spacing");
      categories.add("font-family");
    }
  }
  return categories;
}
function addManifestToken(tokenDef, categoryKey, subGroupName, tokens, byCategory, manifestClassLookup, seen) {
  if (!tokenDef || typeof tokenDef !== "object") return;
  if (!tokenDef.value && !tokenDef.variable && !tokenDef.class) return;
  const varName = tokenDef.variable;
  if (!varName) return;
  const dedupeKey = varName;
  if (seen.has(dedupeKey)) return;
  seen.add(dedupeKey);
  if (FRAMEWORK_INTERNAL_PREFIXES.some((p) => varName.startsWith(p))) return;
  let category = MANIFEST_CATEGORY_MAP[categoryKey] ?? null;
  if (categoryKey === "typography") {
    const nameHint = varName || tokenDef.class || "";
    category = categorizeTypographyToken(nameHint);
  }
  if (!category) return;
  const className = varName ? `var(${varName})` : tokenDef.class || "";
  const ut = {
    className,
    values: varName ? { [varName]: normalizeColorValue(String(tokenDef.value)) } : { _value: normalizeColorValue(String(tokenDef.value)) },
    manifestGroup: subGroupName,
    manifestClass: tokenDef.class || void 0
  };
  tokens.push(ut);
  if (!byCategory.has(category)) byCategory.set(category, []);
  byCategory.get(category).push(ut);
  if (tokenDef.class) {
    manifestClassLookup.set(tokenDef.class, ut);
  }
}
function addManifestTokens(tokens, byCategory, manifestClassLookup) {
  if (!_manifestTokens?.tokens) return;
  const seen = /* @__PURE__ */ new Set();
  for (const [categoryKey, tokenGroup] of Object.entries(_manifestTokens.tokens)) {
    if (!tokenGroup || typeof tokenGroup !== "object") continue;
    for (const [key, entry] of Object.entries(tokenGroup)) {
      if (entry && typeof entry === "object" && (entry.value !== void 0 || entry.variable || entry.class)) {
        addManifestToken(entry, categoryKey, void 0, tokens, byCategory, manifestClassLookup, seen);
      } else if (entry && typeof entry === "object") {
        const subGroupName = key;
        for (const [, subEntry] of Object.entries(entry)) {
          addManifestToken(subEntry, categoryKey, subGroupName, tokens, byCategory, manifestClassLookup, seen);
        }
      }
    }
  }
}
var VAR_CATEGORY_PATTERNS = [
  { pattern: /^--(spacing|space|gap|pad|margin)/i, category: "spacing" },
  { pattern: /^--(size|width|height)/i, category: "sizing" },
  { pattern: /^--(color|bg|text-color|border-color|foreground|background|accent|muted|destructive|primary|secondary)/i, category: "colors" },
  { pattern: /^--(font-size|text-(?:xs|sm|base|lg|xl|\d))/i, category: "font-size" },
  { pattern: /^--(font-weight|font-(?:thin|extralight|light|normal|medium|semibold|bold|extrabold|black))/i, category: "font-weight" },
  { pattern: /^--(leading|line-height)/i, category: "line-height" },
  { pattern: /^--(tracking|letter-spacing|letter)/i, category: "letter-spacing" },
  { pattern: /^--(font-family|font-(?:sans|serif|mono|display|body|heading))/i, category: "font-family" },
  { pattern: /^--(font|text)/i, category: "font-size" },
  { pattern: /^--(radius|border-radius|rounded)/i, category: "border-radius" },
  { pattern: /^--(border-width|border-w|stroke-width)/i, category: "border-width" },
  { pattern: /^--(shadow)/i, category: "box-shadow" },
  { pattern: /^--(opacity|alpha)/i, category: "opacity" }
];
function categoryFromValue(value) {
  const v = value.trim().toLowerCase();
  if (v.startsWith("#") || v.startsWith("rgb") || v.startsWith("hsl") || v.startsWith("oklch") || v.startsWith("oklab")) {
    return "colors";
  }
  const num = parseFloat(v);
  if (!isNaN(num) && /^\d{3}$/.test(v)) {
    return "font-weight";
  }
  if (v.includes(",") || /^(sans-serif|serif|monospace|system-ui|ui-sans-serif|ui-serif|ui-monospace)/.test(v)) {
    return "font-family";
  }
  if (!isNaN(num) && (v.endsWith("px") || v.endsWith("rem") || v.endsWith("em"))) {
    return "spacing";
  }
  if (isSpaceSeparatedRgb(v)) return "colors";
  return null;
}
function categorizeVariable(token) {
  for (const { pattern, category } of VAR_CATEGORY_PATTERNS) {
    if (pattern.test(token.name)) return category;
  }
  return categoryFromValue(token.value);
}
function buildVariableUsageMap() {
  const usageMap = /* @__PURE__ */ new Map();
  function scanRuleList(rules) {
    for (let i = 0; i < rules.length; i++) {
      const rule = rules[i];
      if (rule instanceof CSSGroupingRule || typeof CSSLayerBlockRule !== "undefined" && rule instanceof CSSLayerBlockRule) {
        scanRuleList(rule.cssRules);
        continue;
      }
      if (!(rule instanceof CSSStyleRule)) continue;
      if (!rule.cssText.includes("var(")) continue;
      for (let j = 0; j < rule.style.length; j++) {
        const prop = rule.style.item(j);
        if (prop.startsWith("--")) continue;
        const raw = rule.style.getPropertyValue(prop);
        extractVarUsages(prop, raw, usageMap);
      }
      for (const shorthand of Object.keys(SHORTHAND_LONGHANDS)) {
        const raw = rule.style.getPropertyValue(shorthand);
        if (raw && raw.includes("var(")) {
          for (const lh of SHORTHAND_LONGHANDS[shorthand]) {
            extractVarUsages(lh, raw, usageMap);
          }
        }
      }
    }
  }
  try {
    for (const sheet of document.styleSheets) {
      try {
        scanRuleList(sheet.cssRules);
      } catch {
      }
    }
  } catch {
  }
  return usageMap;
}
function extractVarUsages(prop, raw, map) {
  if (!raw.includes("var(")) return;
  VAR_REF_RE.lastIndex = 0;
  let m;
  while ((m = VAR_REF_RE.exec(raw)) !== null) {
    const varName = m[1];
    if (!map.has(varName)) map.set(varName, /* @__PURE__ */ new Set());
    map.get(varName).add(prop);
  }
}
function categorizeFromUsage(varName, usageMap) {
  const props = usageMap.get(varName);
  if (!props || props.size === 0) return null;
  const counts = /* @__PURE__ */ new Map();
  for (const prop of props) {
    const cat = getCategoryForProperty(prop);
    if (cat) counts.set(cat, (counts.get(cat) ?? 0) + 1);
  }
  if (counts.size === 0) return null;
  let best = null;
  let bestCount = 0;
  for (const [cat, count] of counts) {
    if (count > bestCount) {
      best = cat;
      bestCount = count;
    }
  }
  return best;
}
function getCssVariables() {
  const sheetCount = typeof document !== "undefined" ? document.styleSheets.length : 0;
  if (cssVarCache && cssVarSheetCount === sheetCount) return cssVarCache;
  const tokens = [];
  const byCategory = /* @__PURE__ */ new Map();
  const manifestClassLookup = /* @__PURE__ */ new Map();
  const manifestCategories = getManifestCategories();
  const tokenMap = scanDesignTokens();
  const usageMap = buildVariableUsageMap();
  const seen = /* @__PURE__ */ new Set();
  for (const dt of tokenMap.tokens) {
    if (seen.has(dt.name)) continue;
    seen.add(dt.name);
    if (FRAMEWORK_INTERNAL_PREFIXES.some((p) => dt.name.startsWith(p))) continue;
    const category = categorizeFromUsage(dt.name, usageMap) ?? categorizeVariable(dt);
    if (!category) continue;
    if (manifestCategories.has(category)) continue;
    const ut = {
      className: `var(${dt.name})`,
      values: { [dt.name]: normalizeColorValue(dt.value) }
    };
    tokens.push(ut);
    if (!byCategory.has(category)) byCategory.set(category, []);
    byCategory.get(category).push(ut);
  }
  addManifestTokens(tokens, byCategory, manifestClassLookup);
  cssVarCache = { tokens, byCategory, manifestClassLookup };
  cssVarSheetCount = sheetCount;
  return cssVarCache;
}
function getVariablesForProperty(property) {
  const kebab = camelToKebab(property);
  const category = getCategoryForProperty(kebab);
  if (!category) return [];
  const cssVars = getCssVariables().byCategory.get(category) || [];
  const manifestCategories = getManifestCategories();
  if (manifestCategories.has(category)) return cssVars;
  const registry = getVariableRegistry();
  const classTokens = (registry.groups.get(category) || []).filter((t) => Object.keys(t.values).includes(kebab));
  return [...classTokens, ...cssVars];
}
function hasVariablesForProperty(property) {
  const kebab = camelToKebab(property);
  const category = getCategoryForProperty(kebab);
  if (!category) return false;
  const { byCategory } = getCssVariables();
  if ((byCategory.get(category)?.length ?? 0) > 0) return true;
  if (getManifestCategories().has(category)) return false;
  const registry = getVariableRegistry();
  return (registry.groups.get(category) || []).some((t) => Object.keys(t.values).includes(kebab));
}

// src/utils.ts
function camelToKebab2(prop) {
  if (prop.startsWith("webkit") || prop.startsWith("moz") || prop.startsWith("ms")) {
    return "-" + prop.replace(/([A-Z])/g, "-$1").toLowerCase();
  }
  return prop.replace(/([A-Z])/g, "-$1").toLowerCase();
}
function truncate(str, len) {
  const cleaned = str.replace(/\s+/g, " ").trim();
  return cleaned.length > len ? cleaned.slice(0, len) + "\u2026" : cleaned;
}

// src/inspector/style-source.ts
var LONGHAND_TO_SHORTHAND = {
  "padding-top": "padding",
  "padding-right": "padding",
  "padding-bottom": "padding",
  "padding-left": "padding",
  "margin-top": "margin",
  "margin-right": "margin",
  "margin-bottom": "margin",
  "margin-left": "margin",
  "border-top-width": "border-width",
  "border-right-width": "border-width",
  "border-bottom-width": "border-width",
  "border-left-width": "border-width",
  "border-top-color": "border-color",
  "border-right-color": "border-color",
  "border-bottom-color": "border-color",
  "border-left-color": "border-color",
  "border-top-style": "border-style",
  "border-right-style": "border-style",
  "border-bottom-style": "border-style",
  "border-left-style": "border-style",
  "border-top-left-radius": "border-radius",
  "border-top-right-radius": "border-radius",
  "border-bottom-left-radius": "border-radius",
  "border-bottom-right-radius": "border-radius",
  "gap": "gap",
  "row-gap": "gap",
  "column-gap": "gap"
};
function findStyleSources(element, properties) {
  const result = /* @__PURE__ */ new Map();
  for (const prop of properties) {
    result.set(prop, []);
  }
  const inlineStyle = element.style;
  if (inlineStyle) {
    for (const prop of properties) {
      const kebab = camelToKebab2(prop);
      let value = inlineStyle.getPropertyValue(kebab);
      let matchedProperty = kebab;
      if (!value) {
        const shorthand = LONGHAND_TO_SHORTHAND[kebab];
        if (shorthand) {
          value = inlineStyle.getPropertyValue(shorthand);
          matchedProperty = shorthand;
        }
      }
      if (value) {
        result.get(prop).push({
          property: matchedProperty,
          value: value.trim(),
          selector: "[inline]",
          origin: "inline",
          important: inlineStyle.getPropertyPriority(matchedProperty) === "important"
        });
      }
    }
  }
  for (const sheet of document.styleSheets) {
    try {
      const href = sheet.href || sheet.ownerNode?.textContent?.slice(0, 50) || "embedded";
      walkRules2(sheet.cssRules, element, properties, result, formatSheetName(href));
    } catch {
    }
  }
  for (const [, sources] of result) {
    sources.sort((a, b) => {
      if (a.origin === "inline" && b.origin !== "inline") return -1;
      if (b.origin === "inline" && a.origin !== "inline") return 1;
      if (a.important && !b.important) return -1;
      if (b.important && !a.important) return 1;
      return 0;
    });
  }
  return result;
}
function walkRules2(rules, element, properties, result, sheetName, mediaQuery) {
  for (const rule of rules) {
    if (rule instanceof CSSMediaRule) {
      walkRules2(rule.cssRules, element, properties, result, sheetName, rule.conditionText);
    } else if (rule instanceof CSSStyleRule) {
      try {
        if (!element.matches(rule.selectorText)) continue;
      } catch {
        continue;
      }
      const style = rule.style;
      for (const prop of properties) {
        const kebab = camelToKebab2(prop);
        let value = style.getPropertyValue(kebab);
        let matchedProperty = kebab;
        if (!value) {
          const shorthand = LONGHAND_TO_SHORTHAND[kebab];
          if (shorthand) {
            value = style.getPropertyValue(shorthand);
            matchedProperty = shorthand;
          }
        }
        if (value) {
          const source = {
            property: matchedProperty,
            value: value.trim(),
            selector: rule.selectorText,
            origin: "stylesheet",
            stylesheet: sheetName,
            important: style.getPropertyPriority(matchedProperty) === "important"
          };
          if (mediaQuery) {
            source.mediaQuery = mediaQuery;
          }
          result.get(prop).push(source);
        }
      }
    }
  }
}
function formatSheetName(href) {
  if (href === "embedded") return "embedded <style>";
  try {
    const url = new URL(href);
    const path = url.pathname;
    return path.replace(/^\/_next\/static\/css\//, "").replace(/^\//, "");
  } catch {
    return href.slice(0, 80);
  }
}

// src/engine/candidates.ts
var MAX_ALTERNATIVES = 3;
var MAX_CSS_VARS = 2;
var PSEUDO_STATES = [":hover", ":focus", ":active", ":focus-visible", ":focus-within"];
function stripPseudoState(selector) {
  for (const pseudo of PSEUDO_STATES) {
    if (selector.endsWith(pseudo)) {
      return selector.slice(0, -pseudo.length);
    }
  }
  return selector;
}
function resolvePropertyCandidates(prop, tokenMap, element) {
  const kebab = camelToKebab2(prop.property);
  getVariableRegistry();
  let recommended;
  const utilToken = findVariableForValue(kebab, prop.to);
  if (utilToken) {
    const tokenVal = utilToken.values[kebab] || "";
    const normalizedToken = tokenVal.trim().toLowerCase();
    const normalizedTarget = prop.to.trim().toLowerCase();
    const isExact = normalizedToken === normalizedTarget;
    const candidateType = isRawUtility(utilToken) ? "utility-class" : "semantic-token";
    recommended = {
      type: candidateType,
      name: utilToken.className,
      value: tokenVal,
      exact: isExact,
      distance: isExact ? void 0 : `nearest: ${tokenVal} vs ${prop.to}`
    };
  }
  const SKIP_CSS_VAR_PROPS = /* @__PURE__ */ new Set(["top", "right", "bottom", "left", "transform"]);
  const cssVarMatches = SKIP_CSS_VAR_PROPS.has(kebab) ? [] : findTokensForValue(prop.to, tokenMap);
  const cssVariables = cssVarMatches.slice(0, MAX_CSS_VARS);
  if (!recommended && cssVariables.length > 0) {
    recommended = {
      type: "css-variable",
      name: `var(${cssVariables[0]})`,
      value: prop.to,
      exact: true
    };
  }
  const alternatives = [];
  const altTokens = getAlternativeVariables(kebab, utilToken || void 0);
  for (const alt of altTokens) {
    if (alternatives.length >= MAX_ALTERNATIVES) break;
    const altVal = alt.values[kebab];
    if (!altVal) continue;
    if (recommended && alt.className === recommended.name) continue;
    alternatives.push({
      type: isRawUtility(alt) ? "utility-class" : "semantic-token",
      name: alt.className,
      value: altVal,
      exact: altVal.trim().toLowerCase() === prop.to.trim().toLowerCase()
    });
  }
  let source;
  let conflicts;
  if (element) {
    try {
      const sources = findStyleSources(element, [prop.property]);
      const propSources = sources.get(prop.property) || [];
      if (propSources.length > 0) {
        const winner = propSources[0];
        source = {
          selector: winner.selector,
          origin: winner.origin,
          stylesheet: winner.stylesheet,
          important: winner.important,
          mediaQuery: winner.mediaQuery
        };
        if (propSources.length > 1) {
          conflicts = propSources.slice(1, 4).map((s) => ({
            selector: s.selector,
            value: s.value,
            important: s.important
          }));
        }
      }
    } catch {
    }
  }
  return {
    ...prop,
    recommended,
    alternatives,
    cssVariables,
    source,
    conflicts
  };
}
function enrichPropertyChanges(changes, tokenMap, selector) {
  let element = null;
  try {
    const baseSelector = stripPseudoState(selector);
    element = document.querySelector(baseSelector);
  } catch {
  }
  return changes.filter((c) => !c.property.startsWith("__")).map((c) => resolvePropertyCandidates(c, tokenMap, element));
}

export { camelToKebab2 as camelToKebab, enrichPropertyChanges, getAncestorScopes, getCategoryForProperty, getDirectReactComponent, getDirectReactProps, getDirectReactState, getReactComponentHierarchy, getReactSource, getSelector, getSelectorCandidates, getSharedSelector, getVariableRegistry, getVariablesForProperty, hasVariablesForProperty, isHashedClass, isRawUtility, resolveVariablesForElement, scoreNamePattern, setManifestTokens, setReactState, stripPseudoState, truncate };
