// retune/dist/chunk-THD5EQWV.js
// src/inspector/tokens.ts
function scanDesignTokens() {
  const tokens = [];
  const valueToTokens = /* @__PURE__ */ new Map();
  try {
    for (const sheet of document.styleSheets) {
      try {
        for (const rule of sheet.cssRules) {
          if (rule instanceof CSSStyleRule) {
            extractTokensFromRule(rule, tokens);
          } else if (rule instanceof CSSMediaRule) {
            for (const nested of rule.cssRules) {
              if (nested instanceof CSSStyleRule) {
                extractTokensFromRule(nested, tokens);
              }
            }
          }
        }
      } catch {
      }
    }
  } catch {
  }
  for (const el of [document.documentElement, document.body]) {
    const computed = getComputedStyle(el);
    for (const token of tokens) {
      const resolved = computed.getPropertyValue(token.name).trim();
      if (resolved && resolved !== token.value) {
        token.value = resolved;
      }
    }
  }
  for (const token of tokens) {
    const normalized = normalizeValue(token.value);
    if (!valueToTokens.has(normalized)) {
      valueToTokens.set(normalized, []);
    }
    valueToTokens.get(normalized).push(token.name);
  }
  return { tokens, valueToTokens };
}
function extractTokensFromRule(rule, tokens) {
  const style = rule.style;
  for (let i = 0; i < style.length; i++) {
    const prop = style.item(i);
    if (prop.startsWith("--")) {
      const value = style.getPropertyValue(prop).trim();
      if (value) {
        tokens.push({
          name: prop,
          value,
          source: rule.selectorText
        });
      }
    }
  }
}
function normalizeValue(value) {
  let v = value.trim().toLowerCase();
  v = v.replace(
    /rgba?\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*(?:,\s*[\d.]+\s*)?\)/g,
    (_, r, g, b) => `rgb(${r}, ${g}, ${b})`
  );
  return v;
}
function findTokensForValue(value, tokenMap) {
  const normalized = normalizeValue(value);
  return tokenMap.valueToTokens.get(normalized) || [];
}
function detectStylingApproach(element) {
  const classes = element.className && typeof element.className === "string" ? element.className.trim().split(/\s+/) : [];
  const tailwindPattern = /^-?(?:m|p|w|h|text|bg|border|flex|grid|gap|space|rounded|shadow|opacity|font|leading|tracking|z|inset|top|right|bottom|left|min|max|overflow|cursor|transition|duration|ease|delay|animate|scale|rotate|translate|skew|origin|ring|outline|placeholder|divide|sr|not-sr|container|prose|aspect|columns|break|decoration|underline|overline|line-through|no-underline|antialiased|subpixel|italic|not-italic|normal-case|uppercase|lowercase|capitalize|truncate|indent|align|whitespace|hyphens|content|list|object|overflow|scroll|snap|touch|select|resize|appearance|accent|caret|will-change|fill|stroke)\b/;
  const tailwindCount = classes.filter((c) => tailwindPattern.test(c) || /^(sm|md|lg|xl|2xl|dark|hover|focus|active|group|peer):/.test(c)).length;
  if (tailwindCount >= 3 || classes.length > 0 && tailwindCount / classes.length > 0.5) {
    return "tailwind";
  }
  if (classes.some((c) => /^[a-zA-Z]+_[a-zA-Z]+__[a-zA-Z0-9]+$/.test(c) || /^_[a-zA-Z]+_[a-zA-Z0-9]+$/.test(c))) {
    return "css-modules";
  }
  if (classes.some((c) => /^sc-[a-zA-Z0-9]+$/.test(c) || /^css-[a-zA-Z0-9]+$/.test(c))) {
    return "css-in-js";
  }
  if (element.hasAttribute("data-styled") || element.closest("[data-styled]")) {
    return "styled-components";
  }
  if (classes.length > 0) return "plain-css";
  return "unknown";
}
function summarizeTokenSystem(tokenMap) {
  if (tokenMap.tokens.length === 0) return null;
  const categories = {};
  for (const token of tokenMap.tokens) {
    const prefix = token.name.replace(/^--/, "").split("-").slice(0, 2).join("-");
    categories[prefix] = (categories[prefix] || 0) + 1;
  }
  const sorted = Object.entries(categories).sort((a, b) => b[1] - a[1]).slice(0, 10);
  const total = tokenMap.tokens.length;
  const prefixes = sorted.map(([prefix, count]) => `${prefix} (${count})`).join(", ");
  return `${total} CSS custom properties found. Top prefixes: ${prefixes}`;
}

export { detectStylingApproach, findTokensForValue, scanDesignTokens, summarizeTokenSystem };
