// retune/dist/setup-SMPFR3HY.js
#!/usr/bin/env node
import "./chunk-PDX44BCA.js";

// src/mcp/setup.ts
import { existsSync, mkdirSync, copyFileSync, readFileSync, writeFileSync, readdirSync, statSync } from "fs";
import { join, dirname, extname } from "path";
import { homedir } from "os";
import { fileURLToPath } from "url";
var __filename = fileURLToPath(import.meta.url);
var __dirname = dirname(__filename);
function log(msg) {
  console.log(`[retune] ${msg}`);
}
function warn(msg) {
  console.warn(`[retune] \u26A0 ${msg}`);
}
function findSkillSource() {
  const fromDist = join(__dirname, "..", "skill", "SKILL.md");
  if (existsSync(fromDist)) return dirname(fromDist);
  const fromSrc = join(__dirname, "..", "..", "skill", "SKILL.md");
  if (existsSync(fromSrc)) return dirname(fromSrc);
  return null;
}
function installClaudeCodeSkill(skillSource) {
  const skillDir = join(homedir(), ".claude", "skills", "retune-visual-changes");
  const targetFile = join(skillDir, "SKILL.md");
  const sourceFile = join(skillSource, "SKILL.md");
  try {
    mkdirSync(skillDir, { recursive: true });
    copyFileSync(sourceFile, targetFile);
    log(`Skill installed: ${targetFile}`);
    return true;
  } catch (err) {
    warn(`Could not install Claude Code skill: ${err.message}`);
    return false;
  }
}
function installCursorSkill(skillSource) {
  const skillDir = join(homedir(), ".cursor", "skills", "retune-visual-changes");
  const targetFile = join(skillDir, "SKILL.md");
  const sourceFile = join(skillSource, "SKILL.md");
  try {
    mkdirSync(skillDir, { recursive: true });
    copyFileSync(sourceFile, targetFile);
    log(`Skill installed: ${targetFile}`);
    return true;
  } catch (err) {
    warn(`Could not install Cursor skill: ${err.message}`);
    return false;
  }
}
function setupClaudeCodeMcp() {
  const configPath = join(homedir(), ".claude", "claude_desktop_config.json");
  const configDir = dirname(configPath);
  try {
    mkdirSync(configDir, { recursive: true });
    let config = {};
    if (existsSync(configPath)) {
      config = JSON.parse(readFileSync(configPath, "utf-8"));
    }
    if (!config.mcpServers) config.mcpServers = {};
    config.mcpServers.retune = {
      command: "npx",
      args: ["retune"]
    };
    writeFileSync(configPath, JSON.stringify(config, null, 2));
    log(`MCP configured: ${configPath}`);
    return true;
  } catch (err) {
    warn(`Could not configure Claude Code MCP: ${err.message}`);
    return false;
  }
}
function setupCursorMcp() {
  const configPath = join(homedir(), ".cursor", "mcp.json");
  const configDir = dirname(configPath);
  try {
    mkdirSync(configDir, { recursive: true });
    let config = {};
    if (existsSync(configPath)) {
      config = JSON.parse(readFileSync(configPath, "utf-8"));
    }
    if (!config.mcpServers) config.mcpServers = {};
    config.mcpServers.retune = {
      command: "npx",
      args: ["retune"]
    };
    writeFileSync(configPath, JSON.stringify(config, null, 2));
    log(`MCP configured: ${configPath}`);
    return true;
  } catch (err) {
    warn(`Could not configure Cursor MCP: ${err.message}`);
    return false;
  }
}
function detectTools() {
  const tools = [];
  if (existsSync(join(homedir(), ".claude"))) {
    tools.push("claude-code");
  }
  if (existsSync(join(homedir(), ".cursor"))) {
    tools.push("cursor");
  }
  return tools;
}
function findPublicDir() {
  const cwd = process.cwd();
  const frameworkConfigs = ["next.config.js", "next.config.ts", "next.config.mjs", "vite.config.ts", "vite.config.js", "remix.config.js"];
  const isAppRoot = frameworkConfigs.some((f) => existsSync(join(cwd, f)));
  if (isAppRoot) {
    for (const dir of ["public", "static"]) {
      const path = join(cwd, dir);
      if (existsSync(path) && statSync(path).isDirectory()) return path;
    }
    return join(cwd, "public");
  }
  const appDirs = ["app", "client", "web", "frontend", "packages/app", "packages/web", "apps/web", "apps/client"];
  for (const appDir of appDirs) {
    const appPath = join(cwd, appDir);
    if (!existsSync(appPath)) continue;
    const hasConfig = frameworkConfigs.some((f) => existsSync(join(appPath, f)));
    if (hasConfig) {
      for (const dir of ["public", "static"]) {
        const path = join(appPath, dir);
        if (existsSync(path) && statSync(path).isDirectory()) return path;
      }
      return join(appPath, "public");
    }
  }
  return join(cwd, "public");
}
function extractTokensFromCss() {
  const tokens = {};
  const cssVarRe = /--([\w-]+)\s*:\s*([^;]+)/g;
  const skipPrefixes = ["tw-", "chakra-", "mantine-", "radix-", "nextui-"];
  function scanDir(dir) {
    if (!existsSync(dir)) return;
    try {
      for (const entry of readdirSync(dir)) {
        const fullPath = join(dir, entry);
        try {
          const stat = statSync(fullPath);
          if (stat.isDirectory()) {
            if (["node_modules", ".git", "dist", "build", ".next", ".cache"].includes(entry)) continue;
            scanDir(fullPath);
          } else if ([".css", ".scss", ".less"].includes(extname(entry))) {
            const content = readFileSync(fullPath, "utf-8");
            let match;
            while ((match = cssVarRe.exec(content)) !== null) {
              const name = match[1];
              const value = match[2].trim();
              if (skipPrefixes.some((p) => name.startsWith(p))) continue;
              if (!value || value.startsWith("var(")) continue;
              let category = null;
              if (/^(color|bg|text-color|border-color|foreground|background|accent|brand|primary|secondary|success|danger|warning|error)/i.test(name)) {
                category = "colors";
              } else if (/^(spacing|space|gap|pad|margin)/i.test(name)) {
                category = "spacing";
              } else if (/^(size|width|height)/i.test(name)) {
                category = "sizing";
              } else if (/^(radius|border-radius)/i.test(name)) {
                category = "radii";
              } else if (/^(border-width)/i.test(name)) {
                category = "borderWidths";
              } else if (/^(shadow|elevation)/i.test(name)) {
                category = "shadows";
              } else if (/^(font|text|leading|tracking|letter)/i.test(name)) {
                category = "typography";
              }
              if (!category) continue;
              if (!tokens[category]) tokens[category] = {};
              tokens[category][name] = { value, variable: `--${name}` };
            }
          }
        } catch {
        }
      }
    } catch {
    }
  }
  for (const dir of ["src", "app", "styles", "css", "."]) {
    scanDir(join(process.cwd(), dir === "." ? "" : dir));
  }
  return tokens;
}
function generateManifest() {
  const publicDir = findPublicDir();
  if (!publicDir) return false;
  const manifestPath = join(publicDir, "retune.manifest.json");
  let manifest = {};
  if (existsSync(manifestPath)) {
    try {
      manifest = JSON.parse(readFileSync(manifestPath, "utf-8"));
      if (manifest.tokens) {
        return false;
      }
    } catch {
    }
  }
  const tokens = extractTokensFromCss();
  const tokenCount = Object.values(tokens).reduce((sum, cat) => sum + Object.keys(cat).length, 0);
  if (tokenCount === 0) {
    log("No CSS custom properties found in the project.");
    return false;
  }
  manifest.version = 2;
  manifest.tokens = tokens;
  if (!existsSync(publicDir)) {
    mkdirSync(publicDir, { recursive: true });
  }
  writeFileSync(manifestPath, JSON.stringify(manifest, null, 2));
  log(`Manifest generated: ${manifestPath} (${tokenCount} tokens)`);
  return true;
}
async function setup() {
  log("Setting up Retune...\n");
  const tools = detectTools();
  if (tools.length === 0) {
    warn("No supported AI tools detected (Claude Code, Cursor).");
    log("Install Claude Code or Cursor, then run this command again.");
    return;
  }
  log(`Detected: ${tools.join(", ")}
`);
  const skillSource = findSkillSource();
  if (!skillSource) {
    warn("Could not find skill files. The package may be incomplete.");
  }
  const results = [];
  for (const tool of tools) {
    const result = { tool, mcp: false, skill: false };
    switch (tool) {
      case "claude-code":
        result.mcp = setupClaudeCodeMcp();
        if (skillSource) {
          result.skill = installClaudeCodeSkill(skillSource);
        }
        break;
      case "cursor":
        result.mcp = setupCursorMcp();
        if (skillSource) {
          result.skill = installCursorSkill(skillSource);
        }
        break;
    }
    results.push(result);
  }
  const manifestGenerated = generateManifest();
  console.log("\n--- Setup Summary ---\n");
  for (const r of results) {
    const mcpStatus = r.mcp ? "configured" : "manual setup needed";
    const skillStatus = r.skill ? "installed" : "not installed";
    console.log(`  ${r.tool}: MCP ${mcpStatus}, Skill ${skillStatus}`);
  }
  if (manifestGenerated) {
    console.log(`  manifest: tokens extracted`);
  }
  console.log("");
  if (manifestGenerated) {
    console.log("--- Next Step ---\n");
    console.log("The manifest has design tokens but needs component definitions.");
    console.log("Scan the project's React components and add them to the");
    console.log('"components" section of retune.manifest.json.\n');
    console.log("For each component, include:");
    console.log("  - Props with types (string, number, boolean, enum, function)");
    console.log("  - For enum props, list all allowed values");
    console.log("  - For props that map to CSS classes, include class_map");
    console.log("  - For components with useState, include state hooks");
    console.log("");
    console.log("After updating the manifest, call retune_manifest_loaded");
    console.log("to notify the overlay.\n");
  }
}
export {
  setup
};
